### PR TITLE
Improve performance of long range scans with readahead

### PR DESCRIPTION
**Pull request by sagar0**
This change improves the performance of iterators doing long range scans (e.g. big/full table scans in MyRocks) by using readahead and prefetching additional data on each disk IO. This prefetching is automatically enabled on noticing more than 2 IOs for the same table file during iteration. The readahead size starts with 8KB and is exponentially increased on each additional sequential IO, up to a max of 256 KB. This helps in cutting down the number of IOs needed to complete the range scan. 

Constraints:
- The prefetched data is stored by the OS in page cache. So this currently works only for non direct-reads use-cases i.e applications which use page cache. (Direct-I/O support will be enabled in a later PR). 
- This gets currently enabled only when ReadOptions.readahead_size = 0 (which is the default value).

Thanks to @siying for the original idea and implementation.

**Benchmarks:**
Data fill:
```
TEST_TMPDIR=/data/users/$USER/benchmarks/iter ./db_bench -benchmarks=fillrandom -num=1000000000 -compression_type="none" -level_compaction_dynamic_level_bytes
```
Do a long range scan: Seekrandom with large number of nexts
```
TEST_TMPDIR=/data/users/$USER/benchmarks/iter ./db_bench -benchmarks=seekrandom -duration=60 -num=1000000000 -use_existing_db -seek_nexts=10000 -statistics -histogram
```

Page cache was cleared before each experiment with the command:
```
sudo sh -c "echo 3 > /proc/sys/vm/drop_caches"
```
```
Before:
seekrandom   :   34020.945 micros/op 29 ops/sec;   32.5 MB/s (1636 of 1999 found)
With this change:
seekrandom   :    8726.912 micros/op 114 ops/sec;  126.8 MB/s (5702 of 6999 found)
```
~3.9X performance improvement.

Also verified with strace and gdb that the readahead size is increasing as expected. 
```
strace -e readahead -f -T -t -p <db_bench process pid>
```

### CODE REVIEW

**Comment by siying** [comment]
Please mention in HISTORY.md this improvement and the behavior change.

I still feel like we should give users a way to preserve the previous behavior. Can we do something like that:
```
if (read_options_.readahead_size == 0 || read_options_.readahead_size >= 8 * 1024) {
  // the prefetch logic
  ......
}
```

File: table/block_based_table_reader.cc [hunk 1]
```diff
@@ -1618,6 +1618,19 @@ BlockBasedTable::BlockEntryIteratorState::NewSecondaryIterator(
...
     }
   }
+
+  if (num_file_reads_ < 2) {
+    num_file_reads_++;
+  } else if (handle.offset() + static_cast<size_t>(handle.size()) + kBlockTrailerSize > readahead_limit) {
+    num_file_reads_++;
+    // Do not readahead more than 16MB.
+    readahead_size = std::min(size_t{16 * 1024 * 1024}, readahead_size);
+    table_->rep_->file->Prefetch(handle.offset(), readahead_size);
```
Comment (ajkr) [inline table/block_based_table_reader.cc]: if Prefetch is called a bit before running out of data (so reader never has to wait), do we need such a large readahead size?

Comment (siying) [reply table/block_based_table_reader.cc]: Good point. Maybe 1MB or so is more than enough.

Comment (siying) [reply table/block_based_table_reader.cc]: Maybe 512KB.

Comment (sagar0) [reply table/block_based_table_reader.cc]: I did more experiments and collected data to get a sense for which readahead size gives us the best performance: https://gist.github.com/sagar0/fcba5c2c2d6592929078a295da583d11. Look at the Avg MB/s column (average of 3 runs). Seems like 256KB is the one to go with.

Comment (sagar0) [reply table/block_based_table_reader.cc]: Note that these experiments were run on an SSD. But for HDDs, I am assuming a larger readahead size gives a better performance.

**Comment by sagar0** [comment]
Since we don't have readrandomwhilescanning benchmark implemented yet, I just tested using readrandom and made sure that there is no regression for point lookups. 

```
TEST_TMPDIR=/data/users/$USER/benchmarks/iter ./db_bench -benchmarks=readrandom -duration=60 -num=1000000000 -use_existing_db -statistics -histogram
```
Without this patch:
```
readrandom   :     303.736 micros/op 3289 ops/sec;    0.4 MB/s (194999 of 194999 found)
```
With this patch:
```
readrandom   :     301.907 micros/op 3312 ops/sec;     0.4 MB/s (198999 of 198999 found)
```

**Pull request closed without merging**